warehouse/query-core/src/main/java/datawave/query/jexl/visitors/TreeFlatteningRebuilder.java

```diff
@@ -0,0 +1,328 @@
+            }
+        }
+        
+        return copiedNode;
+    }
+    
+    /**
+     * Returns a copy of the passed in node.
+     *
+     * If the original node has children, those exact children (not copies) will be added to the copied node. However, the parentage of those child nodes will
+     * be left as-is.
+     *
+     * If the original node has no children, we will simply use the RebuildingVisitor to copy the node.
+     *
+     * @param node
+     *            the node to copy
+     * @return the copied node
+     */
+    private JexlNode rebuildNode(JexlNode node) {
+        JexlNode newNode;
+        if (node.jjtGetNumChildren() == 0) {
+            newNode = RebuildingVisitor.copy(node);
+        } else {
+            newNode = JexlNodes.newInstanceOfType(node);
+            newNode.image = node.image;
+            
+            JexlNodes.ensureCapacity(newNode, node.jjtGetNumChildren());
+            
+            int nodeIdx = 0;
+            for (JexlNode child : children(node))
+                newNode.jjtAddChild(child, nodeIdx++);
+        }
+        return newNode;
+    }
+    
+    /**
+     * Determins whether the and node represents a bounded range.
+     *
+     * @param node
+     *            the and node to check
+     * @return true if the and node contains a bounded range, false otherwise
+     */
+    private boolean isBoundedRange(ASTAndNode node) {
+        List<JexlNode> otherNodes = new ArrayList<>();
+        Map<LiteralRange<?>,List<JexlNode>> ranges = JexlASTHelper.getBoundedRangesIndexAgnostic(node, otherNodes, false);
+        return ranges.size() == 1 && otherNodes.isEmpty();
+    }
+    
+    /**
+     * Given a reference node, if we are configured to remove references, we will attempt to remove references and reference expressions, where possible.
+     * Otherwise, we will preserve them.
+     *
+     * @param node
+     *            the reference node to flatten
+     * @return the flattened reference node
+     */
+    private JexlNode flattenReference(ASTReference node) {
+        JexlNode parent = node.jjtGetParent();
+        
+        // if we are told not to remove references OR
+        // if this is a marked node OR
+        // if this is an assignment node OR
+        // if the parent is a NOT node, keep the reference
+        if (!removeReferences || parent instanceof ASTNotNode || JexlASTHelper.dereference(node) instanceof ASTAssignment
+                        || QueryPropertyMarker.instanceOf(node, null)) {
+            return node;
+        }
+        // if this is a (reference -> reference expression) combo with a single child, remove the reference and reference expression
+        else if (node.jjtGetNumChildren() == 1 && node.jjtGetChild(0) instanceof ASTReferenceExpression) {
+            ASTReferenceExpression refExp = (ASTReferenceExpression) node.jjtGetChild(0);
+            if (refExp.jjtGetNumChildren() == 1) {
+                return refExp.jjtGetChild(0);
```

Comment:
Don't we need to fix up the tree here (reset children of parent and reset the parent reference)? I don't see that being done by the caller above.